### PR TITLE
Fixes #3541,#13769 - make sure the ssh orchestration methods return true on success

### PR DESCRIPTION
**Pull request by iNecas**
Strange things start happening otherwise.

### CODE REVIEW

**Comment by iNecas** [comment]
I've noticed similar issues in the cloud-init case as well

**Comment by iNecas** [comment]
I've disabled the `RedundatReturn` cop (it might be just my opinion, but the return has it's place in the code, and the matter of this PR is actually one of those cases)

**Comment by ares** [comment]
+1 for explicit return

**Comment by dLobatog** [comment]
Big -1 to the cop part. It could be enabled just in the orchestration file (enabling it everywhere is crazy), but even there I don't find it'd be appropriate to enable. Return doesn't enforce static typing, it's just a tool in Ruby to be sparingly used for control flow. 

Disabling the cop doesn't mean we enforce any type contracts (we can do that through other mechanisms in the code if it's so important). It introduces yet another thing for PR reviewers to take care of manually, which is (partly) why we use Rubocop, to reduce cognitive load on the reviewers and allow them to focus on behaviour, not syntax.

**Comment by iNecas** [comment]
Could you describe me more, what's wrong about using `return` when it doesn't have to be there? What's the biggest problem of having the explicit return used somewhere where it shouldn't? I feel about this cop as "on of these cops that try you to teach you about some Ruby feature".  I know it doesn't have to be there. I know that sometimes it's better to put it there.

"It introduces yet another thing for PR reviewers to take care of manually, which is (partly) why we use Rubocop, to reduce cognitive load on the reviewers and allow them to focus on behaviour, not syntax." I'm not turning off the whole rubocop, and my argument is that this cop actually participates to behavioral issues in our codebase.

**Comment by ares** [comment]
>  It introduces yet another thing for PR reviewers to take care of manually

I don't think reviewers should check this, I don't see anything negative on being explicit about return. I think all **style** cops are only personal opinions, not necessarily a good practice.

**Comment by dLobatog** [comment]
> Could you describe me more, what's wrong about using return when it doesn't have to be there? What's the biggest problem of having the explicit return used somewhere where it shouldn't? I feel about this cop as "on of these cops that try you to teach you about some Ruby feature". I know it doesn't have to be there. I know that sometimes it's better to put it there.

It's not a big problem, in other projects I've seen people advocating for not returning early, ever. Instead wrap the return value in whatever, and follow the convention of implicit returns. What I see on this PR can be accomplished just by returning `true` everywhere if you want. 

At the core of the issue here is that procedures (method called for their side-effects) are being thought as functions (methods called for their return value). For most of the orchestration code, I'd argue we care about the side effects (`self.errors.present?` when something fails), not return values. 

In fact this issue that's being fixed here (logger returning false halts orchestration) has already been thought of by Rails as inconsistent behavior and [this wouldn't be a bug on Rails 5](http://blog.bigbinary.com/2016/02/13/rails-5-does-not-halt-callback-chain-when-false-is-returned.html), because side-effects is all Rails cares about on validations/callbacks.

> "It introduces yet another thing for PR reviewers to take care of manually, which is (partly) why we use Rubocop, to reduce cognitive load on the reviewers and allow them to focus on behaviour, not syntax." I'm not turning off the whole rubocop, and my argument is that this cop actually participates to behavioral issues in our codebase.

Not sure what you mean by 'participates to behavioral issues'? I'm just saying that while explicit return has a place, turning it on everywhere means reviewers have to check if it's an actual interruption of control flow (good), or if it could be implicit (bad, 30% slower and unexpected for Ruby developers)

In any case [I don't want to bikeshed](http://bikeshed.com/) about this constantly, just want to save me from another check to do on every review. If enough people say yes I guess that's the way forward.

**Comment by iNecas** [comment]
AFAIK the 30% slowness was ruby 1.8.x case only (https://tomafro.net/2009/08/the-cost-of-explicit-returns-in-ruby). I agree that at the core of the issue is the orchestration code checking for the return values, but we can't change that over night without thinking that though. And since we have this behavior there, the explicit return makes sense here IMO.

**Comment by iNecas** [comment]
I wonder if we could have some kind of public team on github for core committers to raise this kind of questions that go beyond the scope of the PR (such as turning on/off rubo cops). Or it already exists and it's just my fault to not find being able to find it?

**Comment by dLobatog** [comment]
@iNecas I think you're looking for @theforeman/core

**Comment by iNecas** [comment]
@dLobatog I'm not sure it works as expected when the the group is `secret`

**Comment by daviddavis** [comment]
I don't have a strong opinion on this but looking at `compute.rb`, I'm confused as to why some methods have just `true` at the end (e.g. match_macs_to_nics) and others have `return true` (e.g. setUserData).

File: app/models/concerns/orchestration/compute.rb [hunk 1]
```diff
@@ -102,6 +103,7 @@ def setUserData
 
     return false if errors.any?
     logger.info "Revoked old certificates and enabled autosign for UserData"
+    return true
```
Comment (stbenjam) [inline app/models/concerns/orchestration/compute.rb:106]: Why would implicit return not work here?  Something like this is more clear to me, and more helpful logs:

``` ruby
if errors.any?
  logger.error "Error revoking old certificates and enabled autosign for UserData: #{errors.full_messages.to_sentence}"
  false
else     
  logger.info "Revoked old certificates and enabled autosign for UserData"
  true
end
```

```
```

File: app/models/concerns/orchestration/compute.rb [hunk 2]
```diff
@@ -74,6 +74,7 @@ def setCompute
     logger.info "Adding Compute instance for #{name}"
     add_interfaces_to_compute_attrs
     self.vm = compute_resource.create_vm compute_attributes.merge(:name => vm_name, :provision_method => provision_method)
+    return true
```
Comment (stbenjam) [inline app/models/concerns/orchestration/compute.rb:77]: Why do we need this? self.vm's truthiness should be a valid return, no?

File: app/models/concerns/orchestration/compute.rb [hunk 3]
```diff
@@ -111,6 +113,7 @@ def delUserData
     if puppetca?
       respond_to?(:initialize_puppetca,true) && initialize_puppetca && delCertificate && delAutosign
     end
+    return true
```
Comment (stbenjam) [inline app/models/concerns/orchestration/compute.rb:116]: Doesn't this potentially hide puppet ca failures?

**Comment by stbenjam** [comment]
Throwing my 2 cents into the pot.

I feel like it's a code smell.  This maybe fixes the specific bugs, but it leaves open the ability to introduce similar ones in the future. As @dLobatog said, you're putting the onus on the reviewer to remember that orchestration methods need these returns.  That's really unfortunate.

I'd rather orchestration be more robust, but like you said it's not going to get rewritten overnight, but I think we can solve it without the explicit returns.  

I think you're even hiding some legitimate errors here.

**Comment by ares** [comment]
@iNecas there are some new questions, also a rebase is needed, moving to WoC

**Comment by bkearney** [comment]
@inecas @ares any chance of bringing this back to life?

**Comment by iNecas** [comment]
@bkearney here you go https://github.com/theforeman/foreman/pull/4262

**Comment by ares** [comment]
@iNecas I don't think this was rejected, I think more votes were :+1: but anyway, thanks for the alternative

**Comment by iNecas** [comment]
At this point, I just need to get the issue resolved; not feeling as getting myself into rubocop fights adventure anytime soon :)

**Pull request closed without merging**